### DISCUSSION TITLE
Best approach to turn a user created folder beneath the virtualization root into a placeholder directory?

### DISCUSSION THREAD

**Discussion by ZabirHoque** [microsoft/ProjFS-Managed-API#75] [Q&A] [answered]
I have a situation where a client creates some directories/files beneath the virtualization root. At a later point, a subset of the directories/files are requested by the client to be backed by the provider's backing store. For files, I'm able to call `PrjDeleteFile()` to virtualize them. Unfortunately, I cannot completely delete folders since they still may contain unbacked files.

I've tried the following, but have not be able to trigger directory enumeration on the user created directories:
- Call `PrjWritePlaceholderInfo()` in response to the user action (not from within any ProjFS callback). Succeeds with `S_OK`, but browsing that directory does not trigger any enumeration callbacks.
- Call `PrjUpdateFileIfNeeded()`. Fails with `ERROR_DIR_NOT_EMPTY` since it still contains some files the user did not mark for backing by the provider. If the directory is empty however (i.e. all files within the directory were marked for backing or the directory is just empty), succeeds with `S_OK` and I receive enumeration callbacks.
- Call `PrjUpdateFileIfNeeded()` during `PRJ_NOTIFY_NEW_FILE_CREATED` to pre-emptively mark all directories as placeholder since they are empty, but fails with `ERROR_SHARING_VIOLATION`.
- Don't call anything specifically, but instead start returning relevant directory as a virtual item during parent directory enumeration. But still won't trigger enumeration of newly backed directory when browsing to it.

I'm very confused how to achieve this scenario. Any guidance would be greatly appreciated. 

Thank you

**Comment by cgallred** [comment, marked as answer]
You want [PrjMarkDirectoryAsPlaceholder](https://docs.microsoft.com/en-us/windows/win32/api/projectedfslib/nf-projectedfslib-prjmarkdirectoryasplaceholder). That will turn a full directory into a placeholder directory, and you should start getting enumeration callbacks on it.

**Reply by ZabirHoque** [reply]
So thats super embarrassing - it worked like a charm. I had dismissed that API since the remarks section made me think it was just for marking virtualization root. Thank you so much!
